---
  title: "hCoCena main KC wt"
author: "Lea Seep"
date: "04 03 2023"
output:
  html_document: default
pdf_document: default
---
  
# Introduction
This Markdown contains the base workflow of hCoCena. You will find detailed information on the satellite functions as well as intermediate outputs that you can use for your custom analyses in the repository's Wiki.

# Pre-Integration Phase

## load hCoCena package
```{r}
library(hcocena)
library(SummarizedExperiment)
```

## create hCoCena-Object

```{r}
init_object()
```


## Working directory setup

  Please set the following parameters to strings that state the path to your expression data, meta data, etc. For more information on the different files and how to pass them to the function, please refer to the following section of the wiki: https://github.com/MarieOestreich/hCoCena/wiki/General-Backgroud-Information#count-files-annotation-files--reference-files
  
  + dir_count_data: path of your count files folder, set FALSE if you are NOT loading expression data from files but instead are using existing 
  data frames in your environment.
  + dir_annotation: path of your annotation files folder, set FALSE if you are NOT loading annotation data from files but instead are using existing 
  data frames in your environment.
  + dir_reference_files: path of your reference files folder (the folder with corresponding files can be downloaded from the github repo https://github.com/MarieOestreich/hCoCena, they are contained in the folder 'reference_files')
  + dir_output: path to the location in which the save folder for this analyses should be created (e.g., "~/" for your home directory).

```{r}

#PATH_TO_REFERENCE_FILES_FOLDER="hCoCena/refFiles/"
PATH_CoCena_results = "hCoCena"


CoCena_Input_KC <- readRDS("../data/CoCena_Input_KC_WT.rds")

vst_data <- counts(CoCena_Input_KC, normalized=TRUE)
#vst_data <- DESeq2::vst(CoCena_Input_KC, blind = F)

#data_seq <- as.data.frame(assay(vst_data))
data_seq <- as.data.frame(vst_data)
annotation_seq <- as.data.frame(colData(CoCena_Input_KC))


init_wd(dir_count_data = F,
        dir_annotation = F,
        dir_reference_files = NULL,
        dir_output = PATH_CoCena_results)

```

  Ensure that all directories exist with the following chunk:

```{r check directories}
check_dirs()
init_save_folder(name = "results_KC_wt")
```

## Defining layers

  For detailed information regarding the structures of the count and annotation data as well as different options for providing data, refer to the function documentation by entering ?hcocena::define_layers into the console.
    
```{r defining Omics Layers}
# Just 1 layer
define_layers(list( RNA_Seq = c("data_seq", "annotation_seq")))


```


  

## Define global settings

  For detailed information regarding the different settings, enter ?hcocena::set_global_settings into the console.

```{r global settings}

set_global_settings(organism = "mouse", 
    								control_keyword = "none", #vs wt_cd_cd_cd
    								variable_of_interest = "Merged", 
    								min_nodes_number_for_network = 50, 
    								min_nodes_number_for_cluster = 50,
    								range_GFC = 2.5,
    								layout_algorithm = "layout_with_fr",
    								data_in_log = F)

```


## Data import

  For detailed information regarding the different parameters, enter ?hcocena::read_data into the console.


```{r data import}
read_data()
```


## OPTIONAL: Data-Based Definition of Top Most Variant Genes

Find inflection points in the ranked variances to filter for the top most variant genes in a data-driven way. 

```{r}
#?? does not work?
hcocena::suggest_topvar()
```

## Define layer-specific settings

  For detailed information regarding the different settings, enter ?hcocena::set_layer_settings into the console.
  

```{r layer-specific settings}

set_layer_settings(top_var = 4000,
                   min_corr = rep(0.60, length(hcobject[["layers"]])), 
                   range_cutoff_length = rep(100, length(hcobject[["layers"]])),
                   print_distribution_plots = rep(F, length(hcobject[["layers"]])))


```

## Data processing part I

  For detailed information on what happens in this step and what parameters can be set, enter ?hcocena::run_expression_analysis_1 into the console.
  

```{r expression analysis up to cutoff}

run_expression_analysis_1(corr_method = "pearson")

```

## Data processing part II


  Choosing the cut-offs:

  Set a correlation cut-off for each of your data sets. To aid this choice, the following plot presents the different cut-off statistics per data set. For more details on cut-offs and the information in this plot as well as the parameters, enter ?hcocena::plot_cutoffs into the console.
  
  
```{r fig.height = 8, fig.width = 15}

plot_cutoffs(interactive = T)

```
  
The correlation cutoff was set for all CoCena analysis the same aiming for a connected network with high correlated edges.

```{r choose cutoff}
set_cutoff(cutoff_vector = c(0.7))

```


  Checking the scale-free topology

  For each data set, the logged degree distribution and the linear regression are plotted to visualize the preservation of the scale-free topology criterion.
  NOTE: Even though biological networks are generally believed to follow a scale-free topology, experience has shown that a lot of transcriptomics data does not follow this principle perfectly. A deviation from the regression line is often observed at higher x-axis values. 


```{r plot degree distribution for chosen cutoff, message = F, warning = F}

plot_deg_dist()

```


  Heatmap of top most variant genes and GFC calculation

  This function plots a heatmap for the network genes in each data layer and computes the Group-Fold-Changes for each genes per layer.


```{r, fig.width = 10, fig.height = 7}

run_expression_analysis_2()

```

# Integration Phase

## Layer integration

  Integrating the layer-specific networks
  
  Here, the previously constructed layer-specific networks will be integrated into one network that combines the information. The integration can be based on the union or intersection of layer-specific networks. Edges that are present in several networks with different lengths can be included based on different options. For detailed information, please refer to the Info Pages in the repository's Wiki.

```{r merge networks}

build_integrated_network(mode = "u", multi_edges = "min")

```



# Post-Integration Phase

## Module detection

Clustering the network

In this step, modules of strong co-expression will be detected in the network and their expression pattern across conditions will be represented in a GFC heatmap. For more information on the available clustering algorithms, run ?hcocena::cluster_calculation and visit the repository's Wiki pages.
  NOTE: You can export your clustering for future use. To do so, please refer to the satellite script, section "Export clustering".
  NOTE 2: Instead of clustering your network here, you can alternatively import a clustering model. To do so, please use the import_clusters() function (see satellite markdown for infos).
  
```{r compute clusters and plot module heatmap}
res=cluster_calculation(cluster_algo = "cluster_leiden",no_of_iterations = 100, max_cluster_count_per_gene = 3) # 1000
#col_anno_categorical(variables = NULL, variable_label = NULL) 
#col_anno_categorical(variables = c("Maternal_diet"), variable_label = "Maternal_diet") 
annotation_seq$Merged <- factor(annotation_seq$Merged,levels = 
                                  c("cdcdcd_Adult",
                                    "cdcdhfd_Adult",
                                    "hfdcdcd_Adult",
                                    "hfdhfdcd_Adult",
                                    "hfdcdhfd_Adult",
                                    "hfdhfdhfd_Adult"),ordered = T)
plot_cluster_heatmap(cluster_columns = F,
                     col_order = levels(annotation_seq$Merged))
```


## OPTIONAL: correlate numeric meta data with clusters

  To see how numeric meta information correlates to the expression patterns of a cluster on a sample basis, refer to the satellite markdown, section "Correlate numeric meta data with clusters".


## Plotting the network coloured by module

NOTE: due to issues in the communication between R and Cytoscape, please refer to the satellite markdown, section "Cytoscape" if you chose Cytoscape as the option for the network layout in your global settings.


```{r plot network coloured by cluster, fig.width=10, fig.height=7}

plot_integrated_network() 


```


## Create gene Lists

  If you would like to evaluate how well each gene belongs to its asserted cluster, please refer to the satellite markdown, section "Cluster scores".
```{r}

KEGGset <- msigdbr::msigdbr(
  species = "Mus musculus",
  category = "C2",
  subcategory = "KEGG"
)
Hallmarkset <- msigdbr::msigdbr(
  species ="Mus musculus",
  category = "H",
)

GOset <- msigdbr::msigdbr(
  species = "Mus musculus",
  category = "C5",
  subcategory = "GO:BP"
)

res <- as.data.frame(res)
clusterGenes <- list()

i = "GLYCOLYSIS"
allTermAssocGenesHallmark <- unique(Hallmarkset[grepl(i,Hallmarkset$gs_name),"ensembl_gene"])
allTermAssocGenesKEGG <- unique(KEGGset[grepl(i,KEGGset$gs_name),"ensembl_gene"])
hifa <- rownames(rowData(CoCena_Input_KC)[which(grepl("Hif1a$",rowData(CoCena_Input_KC)$SYMBOL)),1:3])


for(i in unique(res$color)){
  if(i == "white"){
    next
  }
  clusterGenes[[i]] <- strsplit(res[res$color==i,"gene_n"],split=",")[[1]]
  print(paste0(i, " Hif1a present: ", hifa %in% clusterGenes[[i]] ))

}



#get_cluster_scores()


```

## Highlight GeneSet
```{r}
highlight_geneset_LS<-function (gene_set, col = "black") 
{
  gtc <- GeneToCluster() %>% dplyr::filter(., !color == "white")
  g <- hcobject[["integrated_output"]][["merged_net"]]
  g <- igraph::delete_vertices(g, igraph::V(g)$name[!igraph::V(g)$name %in% 
    gtc$gene])
  base::rownames(gtc) <- gtc$gene
  gtc <- gtc[igraph::V(g)$name, ]
  counter_found <- 0
  igraph::V(g)$color <- base::lapply(gtc$gene, function(x) {
    if (x %in% gene_set) {
      counter_found <- counter_found+1
      dplyr::filter(gtc, gene == x) %>% dplyr::pull(., 
        "color")
    }
    else {
      NA
    }
  }) %>% base::unlist()
  
  print(counter_found)
  igraph::V(g)$size <- base::lapply(gtc$gene, function(x) {
    if (x %in% gene_set) {
      5
    }
    else {
      3
    }
  }) %>% base::unlist()
  igraph::V(g)$frame.color <- base::lapply(gtc$gene, function(x) {
    if (x %in% gene_set) {
      col
    }
    else {
      "lightgrey"
    }
  }) %>% base::unlist()
  l <- hcobject[["integrated_output"]][["cluster_calc"]][["layout"]]
   l <- l[igraph::V(g)$name, ]
  l <- l[base::rownames(l) %in% igraph::V(g)$name, ]
  # Create overlay only
  igraph::E(g)$color <- NA


  igraph::plot.igraph(g, vertex.label = NA, layout = l)
}
gtc <- GeneToCluster() %>% dplyr::filter(., !color == "white")
table(allTermAssocGenesKEGG$ensembl_gene %in% gtc$gene)
highlight_geneset_LS(gene_set = allTermAssocGenesKEGG$ensembl_gene)
```
## Do ORA analysis

```{r GO profiling, fig.width = 10, fig.height = 7, message = F, warning = F}

source("../utils/doORA.R")
library(msigdbr)
library(clusterProfiler)

union_all_genes_in_network <- rownames(hcobject$integrated_output$GFC_all_layers)

input_genes <- rownames(hcobject$layer_specific_outputs$set1$part2$GFC_all_genes)# 5000 gene universe

all_genes <- rownames(vst_data)
# all gene universe
# 


universe_entrez <- clusterProfiler::bitr(input_genes, 
                        fromType="ENSEMBL",
                        toType="ENTREZID",
                        OrgDb="org.Mm.eg.db")$ENTREZID

ORA_cluster_results <- list()
setwd("hCoCena/results_KC_wt/")
for(i in unique(res$color)){ # clusterRes$color
  if(i == "white"){
    next
  }
  fileName = paste0("KC_",i)
  ORA_cluster_results[[i]] <- doOra(
    clusterGenes[[i]], # ENSEBML
    type=c("KEGG","HALLMARK","GO"), #GO
    levelGOTerms=6,
    universe_entrez,
    filename = fileName # will be ORA_[filename][type].png
  )
  
  pval=ORA_cluster_results[[i]]$KEGG[grepl("GLYCOLYSIS",ORA_cluster_results[[i]]$KEGG$ID),"qvalue"]
  pval_H=ORA_cluster_results[[i]]$HALLMARK[grepl("GLYCOLYSIS",ORA_cluster_results[[i]]$HALLMARK$ID),"qvalue"]
  if(length(pval)>0){
      if(pval<0.1){
     print(paste0(i," Gly KEGG: ",pval))
  }
  }
if(length(pval_H)>0){
 if(pval_H<0.1){
   print(paste0(i," Gly HALLMARK: ",pval_H))
 }}
  
  
  pval=ORA_cluster_results[[i]]$KEGG[grepl("HYPOXIA",ORA_cluster_results[[i]]$KEGG$ID),"qvalue"]
  pval_H=ORA_cluster_results[[i]]$HALLMARK[grepl("HYPOXIA",ORA_cluster_results[[i]]$HALLMARK$ID),"qvalue"]
  if(length(pval)>0){
    if(pval<0.1){
     print(paste0(i," Hyp KEGG: ",pval))
    }}
    if(length(pval_H)>0){
   if(pval_H<0.1){
     print(paste0(i," Hyp HALLMARK: ",pval_H))
   }}

  
  pval_H=ORA_cluster_results[[i]]$HALLMARK[grepl("OXIDATIVE",ORA_cluster_results[[i]]$HALLMARK$ID),"qvalue"]
  if(length(pval_H)>0){
    if(pval_H<0.1){
   print(paste0(i," OXIDATIVE HALLMARK: ",pval_H))
 }}
  
  
  
  
  ORA_cluster_results[[i]][["GeneSet"]] <- clusterGenes[[i]]
}



setwd("../../")

## Save for html file of data overview
saveRDS(ORA_cluster_results,"Transcriptomics_ORA_CoCena_results_KC_WT.rds")


```

# get Heatmaps of enriched Pathways

```{r}
selectedPathways <- c("OXIDATIVE_PHOSPHORYLATION","GLYCOLYSIS","HYPOXIA","NoraSet","Oxphos_selected_turq","Glycolysis_selected_plum")
selectedPathways <- c("OXIDATIVE_PHOSPHORYLATION","GLYCOLYSIS")
selectedPathways <- c("Oxphos_selected_turq","Glycolysis_selected_plum","Hypoxia_selected_plum","cytokine_response_steelblue")
selectedPathways <- c("DE_LigandSet")
selectedPathways <- c("OXIDATIVE_PHOSPHORYLATION;turquoise",
                      "OXIDATIVE_PHOSPHORYLATION;lightgreen",
                      "GLYCOLYSIS;steelblue",
                      "cell_activation_involved_in_immune_response;steelblue")
selectedPathways <- c("OXIDATIVE_PHOSPHORYLATION;selected",
                      "GLYCOLYSIS;selected",
                      "cell_activation_involved_in_immune_response;selected")
# subset to those genes of interest w.r.t to cluster
 # i="DEFENSE_RESPONSE_TO_OTHER_ORGANISM"
 # i="RESPONSE_TO_CYTOKINE$"
data2sum <-  vst_data

for(i_all in selectedPathways){
  i_color <- unlist(strsplit(i_all,";"))[2]
  i <- unlist(strsplit(i_all,";"))[1]
  # get all terms KEGG & GO
  print(unique(KEGGset$gs_name[grepl(i,KEGGset$gs_name)]))
  allTermAssocGenesKEGG <- unique(KEGGset[grepl(i,KEGGset$gs_name),"ensembl_gene"])
  unionKEGG <- allTermAssocGenesKEGG$ensembl_gene
  
  
  print(unique(Hallmarkset$gs_name[grepl(i,Hallmarkset$gs_name)]))
  allTermAssocGenesHallmark <- unique(Hallmarkset[grepl(i,Hallmarkset$gs_name),"ensembl_gene"])
  unionHALLMARK <- unique(c(allTermAssocGenesHallmark$ensembl_gene))
  
  
  print(unique(GOset$gs_name[grepl(paste0("GOBP_",i,"$"),GOset$gs_name,ignore.case = T)]))
  allTermAssocGenesGO <- unique(GOset[grepl(paste0("GOBP_",i,"$"),GOset$gs_name,ignore.case = T),"ensembl_gene"])
  
  if(i_color!="selected"){
      allTermAssocGenesGO <- allTermAssocGenesGO$ensembl_gene
  unionGO <- unique(c(allTermAssocGenesGO))
  
  
  #select all genes present in cluster of interest
  if(i=="OXIDATIVE_PHOSPHORYLATION"){
     union_selected <- unionKEGG[unionKEGG %in% ORA_cluster_results[[i_color]]$GeneSet]
      sort(rowData(CoCena_Input_KC)[union_selected,"SYMBOL"])
  }

  if(i =="GLYCOLYSIS"){
     union_selected <- unionKEGG[unionKEGG %in% ORA_cluster_results[[i_color]]$GeneSet]
     # to tranlsate
    union_trans <- rowData(CoCena_Input_KC)[union_selected,"SYMBOL"]
    paste(union_trans,collapse=",")
     
  }
    if(i =="cell_activation_involved_in_immune_response"){
     union_selected <- unionGO[unionGO %in% ORA_cluster_results[[i_color]]$GeneSet]
     union_trans <- rowData(CoCena_Input_KC)[union_selected,"SYMBOL"]
    }
  }else{
      if(i == "OXIDATIVE_PHOSPHORYLATION"){
        union_trans <- c("Ndufv1","Uqcrfs1","Sdha","Uqcrh","Cox4i1",
                         "Ndufs3","Atp5e","Atp6v0b","Atp6v0e","Ndufa5",
                         "Atp5j","Sdhb","Atp5o","Ndufb7","Ndufb8","Ndufa8",
                         "Ndufb10","Cox17","Atp5j2","Ndufc1","Atp5k","Ndufa7",
                         "Ndufb6","Uqcrq","Cox5b","Atp6v1g1","Cox7c","Ndufa9","Ndufa4")
        union_selected <- rownames(rowData(CoCena_Input_KC)[rowData(CoCena_Input_KC)[,"SYMBOL"]%in%union_trans,])
      }
      if(i == "GLYCOLYSIS"){
        union_trans <- c("Aldh2","Hk3","Aldh3b1","Adh5","Pfkl","Pgam1",
                         "Tpi1","Pkm","Aldoa","Ldha","Gpi1","Gapdh","Eno1")
        union_selected <- rownames(rowData(CoCena_Input_KC)[rowData(CoCena_Input_KC)[,"SYMBOL"]%in%union_trans,])
      }
      if(i == "cell_activation_involved_in_immune_response"){
        union_trans <- c("Ptk2b","Lfng","Irf8","Il4ra","Slc11a1","Lrp1",
                         "Il6ra","Itgal","Adgre5","Abl1","Tyrobp","Btk",
                         "Hmox1","Relb","Spi1","Sphk2","Pik3cd","Lgals9",
                         "Stat6","Lgals3","Fgr","Itgam","Syk","Rac2","Icam1",
                         "Fcgr2b","Ccr2","Lgals1","Cx3cr1","Trem2","Myd88",
                         "Trp53","Grn","Fes","Gpnmb")
        union_selected <- rownames(rowData(CoCena_Input_KC)[rowData(CoCena_Input_KC)[,"SYMBOL"]%in%union_trans,])
      }
    
  }

  print(length(union_selected))
  
  subsetData <- data2sum[rownames(data2sum) %in% union_selected,]
  subsetData_df <- as.data.frame(t(subsetData))
  subsetData_df <- cbind(subsetData_df,colData(CoCena_Input_KC)[rownames(subsetData_df),"Merged"])
  colnames(subsetData_df)[ncol(subsetData_df)] <- "Merged"
  
  
  mean_per_condition = aggregate(x = subsetData_df[,-ncol(subsetData_df)],
                                 by = list(as.factor(subsetData_df[,"Merged"])),
                                 FUN = mean)
  

  rownames(mean_per_condition) <- gsub("_Adult","",mean_per_condition$Group.1)
  mean_per_condition$Group.1 <- NULL
  
  
  
  colnames(mean_per_condition) <- rowData(CoCena_Input_KC)[rownames(rowData(CoCena_Input_KC)) %in% colnames(mean_per_condition),"SYMBOL"]
  
  
  #z_values <- as.data.frame(t(apply(mean_per_condition,1,scale)))
  z_values <- mean_per_condition
  colnames(z_values) <- colnames(mean_per_condition)
  nBins_each <- 5
  myColor <- c(colorRampPalette(c("#0077b6", "#d9effa"))(nBins_each),
               "white",
               colorRampPalette(c("#ffe8e8","#D62828"))(nBins_each)
  )
  
  colorTheme <- c("#c6c6c6","#606060","#c32b38","#fee2d1","#8a4094","#ebdeec")
  names(colorTheme) <- c("cdcdcd",
                        "cdcdhfd",
                        "hfdcdcd",
                        "hfdhfdcd",
                        "hfdcdhfd",
                        "hfdhfdhfd")
  
  annoCol <- list(group = colorTheme)
  col_anno = data.frame(group = names(colorTheme))
  rownames(col_anno) = col_anno$group
  x <- z_values[names(colorTheme),] 

  
  savingName = paste0("hCoCena/results_KC_wt/Heatmap_",i_color,"_",i,".svg")
svg(savingName)
  pheatmap(x,
                 color = myColor,
                 scale="column",
                # breaks = myBreaks, # no manual adjustment needed if scaling 
                # display_numbers = as.data.frame(t(result_sig)), # needs to be a matrix of same dim as LFC Table
                 fontsize_number = 15,
                 cellheight = 10,
                 cellwidth = 10,
                #filename = givenFilename,
                 annotation_colors = annoCol,
                 annotation_row  = col_anno,
                 treeheight_row = 0,
                 treeheight_col = 0,
                main = i,
                angle_col = 90,
                cluster_rows = F
  )
  dev.off()
}


```

